File: src/main/java/io/github/thejaxonhill/tarotapi/usecase/DrawTarotCard.java
package io.github.thejaxonhill.tarotapi.usecase;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import org.springframework.stereotype.Service;

import io.github.thejaxonhill.tarotapi.TarotCard;
import lombok.RequiredArgsConstructor;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
@RequiredArgsConstructor
public class DrawTarotCard {

    private final List<TarotCard> tarotCards;

    public Mono<TarotCard> drawCard(List<Integer> alreadyDrawnIds) {
        return Mono.from(drawCards(1L, alreadyDrawnIds));
    }

    public Flux<TarotCard> drawCards(Long amountToDraw, List<Integer> alreadyDrawnIds) {
        List<TarotCard> deck = new ArrayList<>(tarotCards.stream()
                .filter(card -> alreadyDrawnIds == null || !alreadyDrawnIds.contains(card.getId()))
                .toList());
        Collections.shuffle(deck);

        return Flux.fromStream(deck.stream().limit(amountToDraw != null && amountToDraw > 0 ? amountToDraw : 3L));
    }

}

File: src/main/kotlin/io/github/thejaxonhill/tarotapi/application/usecase/DrawTarotCard.kt
package io.github.thejaxonhill.tarotapi.application.usecase

import io.github.thejaxonhill.tarotapi.application.stereotype.UseCase
import io.github.thejaxonhill.tarotapi.domain.TarotCard
import io.github.thejaxonhill.tarotapi.domain.TarotCardRepository

@UseCase
class DrawTarotCard(private val tarotCardRepository: TarotCardRepository) {
    fun draw(alreadyDrawnIds: List<Int> = emptyList()): TarotCard = tarotCardRepository.loadAll()
            .filter { !alreadyDrawnIds.contains(it.id) }
            .randomOrNull() ?: throw IllegalStateException("No cards left to draw")
}
